src/net_processing.cpp

```diff
@@ -1981,15 +1981,21 @@ void static ProcessOrphanTx(CConnman* connman, CTxMemPool& mempool, std::set<uin
 }
 
 /**
- *  A block has been processed. Do housekeeping.
+ *  A block has been processed. Handle potential peer punishment and housekeeping.
  */
-void static BlockProcessed(CNode* pfrom, std::shared_ptr<CBlock> pblock, bool new_block)
+void static BlockProcessed(CNode* pfrom, CConnman* connman, std::shared_ptr<CBlock> pblock, BlockValidationState& state, bool new_block)
 {
-    if (new_block) {
-        pfrom->nLastBlockTime = GetTime();
-    } else {
+    if (!state.IsValid()) {
+        // The block failed anti-dos / mutation checks. Call BlockChecked() callback here.
+        // This clears the block from mapBlockSource.
+        BlockChecked(*pblock, state, connman);
+    } else if (!new_block) {
```

Comment:
8fc192c

How does this new check behaves with regards to write-failure, first-seen block ? In `AcceptBlock`, if `blockPos` returned by `SaveBlockToDisk` is null, an error is set and not an invalid state, so it won't be erased by above `BlockChecked` call. I think it's catch up in current code but not after changes.